Feat: RWA Token Factory

nduchak opened the pull request:
[sc-453]

nduchak committed:
feat: add rwa token factory

nduchak committed:
doc: update README

nduchak committed:
dapp install https://github.com/hbarcelos/forward-proxy

nduchak committed:
chore: adjust remapping

hbarcelos commented on README.md:
```diff
@@ -10,6 +10,7 @@
 - `RwaInputConduit`: repays DAI.
 - `RwaInputConduit2`: variation of `RwaInputConduit` with an whitelist to control permissions to repay DAI.
 - `RwaToken`: represents the RWA collateral in the system.
+- `RwaTokenFactory`: factory for deploying RwaToken.
```
Nitpicking here, but maybe "factory of `RwaToken`s" is a more appropriate way to write this.

```suggestion
- `RwaTokenFactory`: factory of `RwaToken`s.
```

hbarcelos commented on src/invariants/RwaTokenInvariant.sol:
```diff
@@ -22,7 +22,7 @@
 pragma solidity 0.6.12;
 
-import "../tokens/RwaToken.sol";
+import "../tokens/rwa/RwaToken.sol";
```
I wonder if we need this subdirectory... the entire repo is for RWA, so maybe let's just have it within `./tokens/RwaToken.sol`?

nduchak replied:
Just though we have a `tokens` folder which means we can have multiple tokens and this factory will work only for the current one.

nduchak replied:
In case we have more tokens maybe then we can move `Factory` somewhere else???

hbarcelos reviewed (commented):
LGTM overall, just some comments regarding the structure.

This is my personal opinion, but I usually like to avoid deeply nested directory structure, because it makes it harder to find files.

I don't think we're going to have anything not related o RWA in this repo, so maybe a `src/tokens/rwa` directory is overkill and we could keep those files within `src/tokens/` instead.

nduchak committed:
Update README.md

Co-authored-by: Henrique Barcelos <16565602+hbarcelos@users.noreply.github.com>

nduchak committed:
chore: remove revert structure change for tokens die

nduchak committed:
Merge remote-tracking branch 'origin/feat/rwa-token-factory' into feat/rwa-token-factory

hbarcelos commented on src/tokens/RwaToken.t.sol:
```diff
@@ -0,0 +1,54 @@
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity >=0.6.8 <0.7.0;
```
We should probably start using a fixed compiler version here:

```suggestion
pragma solidity 0.6.12;
```

hbarcelos commented on src/tokens/RwaTokenFactory.sol:
```diff
@@ -0,0 +1,102 @@
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity >=0.6.8 <0.7.0;
```
Same as above:
```suggestion
pragma solidity 0.6.12;
```

hbarcelos commented on src/tokens/RwaTokenFactory.sol:
```diff
@@ -0,0 +1,102 @@
+
+contract RwaTokenFactory {
+    // --- registry ---
+    mapping (bytes32 => bool) public tokensData;
```
I'm trying to think about other data that we might want to store here instead of having just a boolean... Not sure yet.

nduchak replied:
Just add a token address

hbarcelos commented on src/tokens/RwaTokenFactory.t.sol:
```diff
@@ -0,0 +1,88 @@
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity >=0.6.8 <0.7.0;
```
Same as above:

```suggestion
pragma solidity 0.6.12;
```

hbarcelos commented on src/tokens/RwaTokenFactory.sol:
```diff
@@ -0,0 +1,102 @@
+        return tokens;
+    }
+
+    function createRwaToken(string calldata name, string calldata symbol, address recipient) public auth  returns (RwaToken) {
```
Usually `external` is preferred over `public` when the method is not accessed internally, but as of solidity 0.6.9 it doesn't have any difference in the bytecode. However, for humans reading the code it might matter.

```suggestion
    function createRwaToken(string calldata name, string calldata symbol, address recipient) external auth returns (RwaToken) {
```

hbarcelos commented on src/tokens/RwaTokenFactory.t.sol:
```diff
@@ -0,0 +1,88 @@
+/* prettier-disable */
+
+/**
+ * This file is a copy of https://goerli.etherscan.io/address/0xeb7C7DE82c3b05BD4059f11aE8f43dD7f1595bce#code.
+ * The only change is the solidity version, since this repo is using 0.6.x
+ */
+
+////// src/RwaToken.sol
```
These comments should probably be removed and/or replaced.
Since it's a new file, it does not make sense to disable prettier, for example.

hbarcelos replied:
```suggestion
```

hbarcelos commented on src/tokens/RwaTokenFactory.t.sol:
```diff
@@ -0,0 +1,88 @@
+ */
+
+////// src/RwaToken.sol
+// Copyright (C) 2020, 2021 Lev Livnev <lev@liv.nev.org.uk>
```
We should probably remove Lev's name and list only the Dai Foundation here, since it's a new contract.

hbarcelos replied:
```suggestion
// Copyright (C) 2022 Dai Foundation
```

hbarcelos commented on src/tokens/RwaTokenFactory.sol:
```diff
@@ -0,0 +1,102 @@
+/* prettier-disable */
+
+/**
+ * This file is a copy of https://goerli.etherscan.io/address/0xeb7C7DE82c3b05BD4059f11aE8f43dD7f1595bce#code.
+ * The only change is the solidity version, since this repo is using 0.6.x
+ */
+
```
These comments are not relevant anymore.

hbarcelos commented on src/tokens/RwaTokenFactory.sol:
```diff
@@ -0,0 +1,102 @@
+ * The only change is the solidity version, since this repo is using 0.6.x
+ */
+
+////// src/RwaToken.sol
```
I think it's redundant to have the file name in a comment.
```suggestion
```

hbarcelos commented on src/tokens/RwaTokenFactory.sol:
```diff
@@ -0,0 +1,102 @@
+ */
+
+////// src/RwaToken.sol
+// Copyright (C) 2020, 2021 Lev Livnev <lev@liv.nev.org.uk>
```
```suggestion
// Copyright (C) 2022 Dai Foundation
```

hbarcelos commented on src/tokens/RwaTokenFactory.t.sol:
```diff
@@ -0,0 +1,88 @@
+        tokenFactory.createRwaToken(name, symbol, address(this));
+    }
+
+    function testFail_NameAndSymbolRequired() public {
```
```suggestion
    function testFail_nameAndSymbolRequired() public {
```

hbarcelos commented on src/tokens/RwaTokenFactory.t.sol:
```diff
@@ -0,0 +1,88 @@
+        RwaTokenFactory(op._(address(tokenFactory))).createRwaToken("", "", address(this));
+    }
+
+    function testFail_RecipientRequired() public {
```
```suggestion
    function testFail_recipientRequired() public {
```

hbarcelos reviewed (commented):
Just some nitpicking with the comments. It's looking good.

1. I'm only not sure whether we don't want to have any other data regarding the token in the `tokensData` mapping. If nothing else should be added, then maybe this variable should have a different name, like `tokenRegistry` or similar.
2. We should probably have an event for when a new `RwaToken` is created:
    ```solidity
    event RwaTokenCreated(string name, string indexed symbol, address indexed recipient);
    ```
    Probably we don't need to index `name`, but `symbol` and `address` might be useful for external parties trying to build tooling around it.

hbarcelos commented on src/tokens/RwaToken.t.sol:
```diff
@@ -0,0 +1,54 @@
+ * The only change is the solidity version, since this repo is using 0.6.x
+ */
+
+////// src/RwaToken.sol
```
```suggestion
```

hbarcelos commented on src/tokens/RwaTokenFactory.sol:
```diff
@@ -0,0 +1,102 @@
+/* prettier-disable */
```
Oh, this is important: since it's a new file, we should probably use NatSpec to document it.

nduchak committed:
feat: add TokenCreated event
store token address in registry

nduchak committed:
chore: fix PR comments

hbarcelos commented on src/tokens/RwaTokenFactory.sol:
```diff
@@ -0,0 +1,100 @@
+/* prettier-disable */
```
```suggestion
```
Please remove this.
We should let `prettier` do it's thing on new files.

hbarcelos commented on src/tokens/RwaToken.t.sol:
```diff
@@ -0,0 +1,53 @@
+/* prettier-disable */
```
```suggestion
```

hbarcelos commented on src/tokens/RwaToken.t.sol:
```diff
@@ -0,0 +1,53 @@
+/* prettier-disable */
+
```
```suggestion
```

hbarcelos reviewed (commented):
What are your thoughts on NatSpec for the `RwaTokenFactory.sol` file? Should we just not use it?

nduchak committed:
chore: Add NatSpec docs
restructure SC layout to one proposed by Henrique

nduchak committed:
chore: fix more PR comments

hbarcelos reviewed (commented):
✨ Looking good! :shipit:

My last nitpick is regarding the `tokensData` variable. I'm not happy with this name, but I'm also not sure what would be appropriate here? Maybe `tokenExists`, or `hasToken`?

hbarcelos commented on src/tokens/RwaTokenFactory.sol:
```diff
@@ -0,0 +1,145 @@
+     * @param dsPause DsPause contract address.
+     */
+    constructor(
+        address dsPause
```
Also maybe we shouldn't call this argument `dsPause` as it's an implementation detail. I think `owner` would suffice. In the docs we can make clear that this is meant to be the `DSPauseProxy` in the current MCD incarnation, but could be any other component that has root access to the system.

hbarcelos commented on src/tokens/RwaTokenFactory.sol:
```diff
@@ -0,0 +1,145 @@
+        require(recipient != address(0), "RwaTokenFactory/recipient-not-set");
+        require(bytes(name).length != 0, "RwaTokenFactory/name-not-set");
+        require(bytes(symbol).length != 0, "RwaTokenFactory/symbol-not-set");
+        bytes32 _symbol = this.stringToBytes32(symbol);
```
If I'm not mistaken, using `this.<method>` uses the `CALL` opcode, which is more expensive in terms of gas. I think you can simply call `stringToBytes32(symbol)` without the prefix.

hbarcelos replied:
```suggestion
        bytes32 _symbol = stringToBytes32(symbol);
```

hbarcelos commented on src/tokens/RwaTokenFactory.sol:
```diff
@@ -0,0 +1,145 @@
+     * @notice Helper function for converting string to bytes32
+     * @param source String to convert.
+     */
+    function stringToBytes32(string memory source) external pure returns (bytes32 result) {
```
```suggestion
    function stringToBytes32(string memory source) public pure returns (bytes32 result) {
```

hbarcelos commented on src/tokens/RwaTokenFactory.sol:
```diff
@@ -0,0 +1,145 @@
+        require(dsPause != address(0), "RwaTokenFactory/ds-pause-not-set");
+        wards[dsPause] = 1;
+        emit Rely(dsPause);
+    }
```
```suggestion
    /**
     * @notice Gives `owner` admin access.
     * @dev `onwer` in MCD context, owner is meant to be the DSPauseProxy which has full control of the system.
     * @param owner The owner address.
     */
    constructor(
        address owner
    ) public {
        require(owner != address(0), "RwaTokenFactory/owner-not-set");
        wards[owner] = 1;
        emit Rely(owner);
    }
```

hbarcelos committed:
refactor: fix reentrancy issue

hbarcelos committed:
refactor: rename owner variable

hbarcelos committed:
docs: improve @notice for count() and list()

hbarcelos committed:
refactor: improve string conversion to bytes32

- [x] Change `stringToBytes32` visibility from `external` to `public`.
    Since this method is called internally, if it's external, it means
    it can only be called as `this.stringToBytes32()`, which will use
    the `CALL` opcode, wasting gas.
- [x] Change `source` param type from `string memory` to `string
  calldata`.
    `string memory` input parameter means that whatever is passed as an
    argument is copied to memory, however we are already making such
    copy when casting it to `bytes memory`.
    Using `string calldata` eliminate the additional copy, saving some
    gas.
